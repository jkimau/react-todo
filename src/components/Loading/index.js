import React from 'react';
import styled from 'styled-components';

const LoadingDiv = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -9px;
  margin-left: -35px;
`;

const Loading = () => (
  <LoadingDiv>Loading...</LoadingDiv>
);

export default Loading;
