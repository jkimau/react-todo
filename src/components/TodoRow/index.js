import React, { Component } from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import styled from 'styled-components';

import { mainBG, todoRowBorder } from 'global/colors';

const padding = `15px`;

const TodoRowWrapper = styled.div`
  position: relative;
  background: white;
  border-bottom: 1px solid ${todoRowBorder};
  color: ${props => props.completed ? '#ccc' : 'inherit'};

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    background: ${mainBG};
  }
`;

const Header = styled.div`
  padding: ${padding}
`;

const Body = styled.div`
  padding: ${padding};
`;

const OpenClose = styled.span`
  float: right;
  cursor: pointer;
`;

const TodoRowMenu = styled.ul`
  position: absolute;
  top: 45px;
  right: -100px;
  z-index: 100;
  margin: 0;
  width: 160px;
  border-radius: 4px;
  list-style: none;
  padding-left: 0;
  background: white;
  color: #666;
  overflow: hidden;

  &.slide-enter {
    opacity: 0;
    height: 0px;
    border: none;
    transition: all 0.05s ease;
  }

  &.slide-leave {
    opacity: 1;
    height: ${props => props.totalHeight ? `${props.totalHeight}px` : '100px'};
    border: 1px solid #ddd;
    transition: all 0.05s ease;
  }

  &:focus {
    outline: none;
  }

  li {
    height: ${props => props.listHeight ? `${props.listHeight}px` : 'auto'};
    padding: 6px;
    box-sizing: border-box;
    text-align: center;
    font-size: 14px;
    cursor: pointer;

    &:hover {
      background: #2FC2EF;
      color: white;
    }
  }
`;

export default class TodoRow extends Component {
  constructor(props) {
    super(props);

    this.toggleMenu = this.toggleMenu.bind(this);
  }

  toggleMenu(e) {
    const { todo: { id }, toggleRowMenuHandler } = this.props;
    toggleRowMenuHandler(id);
  }

  render() {
    const {
      todo: {
        title,
        date,
        details,
        completed,
        menuOpen
      }
    } = this.props;
    return (
      <TodoRowWrapper completed={completed}>
        <Header>
          <span>{title}</span>
          <span>{date}</span>
          <OpenClose onClick={this.toggleMenu} className="todo-menu-trigger">
            {menuOpen ? 'close' : 'open'}
          </OpenClose>
          <ReactCSSTransitionGroup
            transitionName="slide"
            transitionEnterTimeout={0}
            transitionLeaveTimeout={0}>
            <TodoRowMenu
              className={menuOpen ? 'slide-leave' : 'slide-enter'}
              totalHeight={30 * 2}
              listHeight={30}>
              <li>Mark complete</li>
              <li>Delete</li>
            </TodoRowMenu>
          </ReactCSSTransitionGroup>
        </Header>
        <Body>{details}</Body>
      </TodoRowWrapper>
    );
  }
}
