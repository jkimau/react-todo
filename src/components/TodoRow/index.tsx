import React, { Component } from 'react';
import styled from 'styled-components';

import { mainBG, todoRowBorder } from 'global/colors';
import { TodoPropTypes } from 'propTypes';

const padding = `15px`;

const TodoRowWrapper = styled.div`
  position: relative;
  background: white;
  border-bottom: 1px solid ${todoRowBorder};
  color: ${props => props.completed ? '#ccc' : 'inherit'};

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    background: ${mainBG};
  }
`;

const Header = styled.div`
  padding: ${padding}
`;

const Body = styled.div`
  padding: ${padding};
`;

const OpenClose = styled.span`
  float: right;
  cursor: pointer;
`;

const TodoRowMenu = styled.ul`
  position: absolute;
  top: 45px;
  right: -100px;
  z-index: 100;
  margin: 0;
  width: 160px;
  border-radius: 4px;
  list-style: none;
  padding-left: 0;
  background: white;
  color: #666;
  overflow: hidden;

  &.slide-enter {
    opacity: 0;
    height: 0px;
    border: none;
    transition: all 0.05s ease;
  }

  &.slide-leave {
    opacity: 1;
    height: ${props => props.totalHeight ? `${props.totalHeight}px` : '100px'};
    border: 1px solid #ddd;
    transition: all 0.05s ease;
  }

  &:focus {
    outline: none;
  }

  li {
    height: ${props => props.listHeight ? `${props.listHeight}px` : 'auto'};
    padding: 6px;
    box-sizing: border-box;
    text-align: center;
    font-size: 14px;
    cursor: pointer;

    &:hover {
      background: #2FC2EF;
      color: white;
    }
  }
`;

interface TodoRowPropTypes {
  key: number,
  todo: TodoPropTypes,
  toggleTodoMenu(id: number): void
};

export default class TodoRow extends Component<TodoRowPropTypes, null> {
  constructor(props) {
    super(props);

    this.toggleTodoMenu = this.toggleTodoMenu.bind(this);
    this.stopPropagateEvent = this.stopPropagateEvent.bind(this);
  }

  toggleTodoMenu() {
    this.props.toggleTodoMenu(this.props.todo.id);
  }

  stopPropagateEvent(e) {
    e.stopPropagation();
  }

  render() {
    const {
      todo: {
        title,
        date,
        details,
        completed,
        menuOpen
      }
    } = this.props;

    return (
      <TodoRowWrapper completed={completed}>
        <Header>
          <span className="todo-row-header">{title}</span>
          <span className="todo-row-date">{date}</span>
          <OpenClose onClick={this.toggleTodoMenu} className="todo-menu-trigger">
            {menuOpen ? 'close' : 'open'}
          </OpenClose>
          <TodoRowMenu
            onClick={this.stopPropagateEvent}
            className={`todo-row-menu ${menuOpen ? 'slide-leave' : 'slide-enter'}`}
            totalHeight={30 * 2}
            listHeight={30}>
            <li>Mark complete</li>
            <li>Delete</li>
          </TodoRowMenu>
        </Header>
        <Body className="todo-row-details">{details}</Body>
      </TodoRowWrapper>
    );
  }
}
