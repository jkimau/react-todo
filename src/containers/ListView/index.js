import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import TodoRow from 'components/TodoRow';
import Loading from 'components/Loading';
import { todoRowBorder } from 'global/colors';
import { fetchTodos, setViewMode, toggleTodoMenu } from 'actions';

const ListViewWrapper = styled.div`
  height: 100%;
  padding-top: 40px;
`;

const ContentContainer = styled.div`
  min-width: 400px;
  max-width: 590px;
  padding: 0 20px;
  margin: 0 auto;
  box-sizing: border-box;
`;

const TodoListWarpper = styled.div`
  border: 1px solid ${todoRowBorder};
  margin-bottom: 80px;
`;

export class ListView extends Component {
  componentDidMount() {
    if (!this.props.isFetching) {
      this.props.fetchTodos();
    }
  }

  shouldComponentUpdate(nextState) {
    const hasSameTodos = this.props.todos === nextState.todos;
    return !hasSameTodos;
  }

  render() {
    const todoListWrapper = () => (
      <TodoListWarpper>
        {this.props.todos.map(todo => (
          <TodoRow
            key={todo.id}
            todo={todo}
            toggleTodoMenu={this.props.toggleTodoMenu}
          />
        ))}
      </TodoListWarpper>
    );

    const loading = () => (<Loading/>);

    return (
      <ListViewWrapper>
        <h1>List view page</h1>
        <ContentContainer>
          {
            this.props.todos
            && this.props.todos.length > 0
            ? todoListWrapper()
            : loading()
          }
        </ContentContainer>
      </ListViewWrapper>
    );
  }
}

const mapStateToProps = ({ todoState: { isFetching, todos }}) => {
  return { isFetching, todos }
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTodos: () => dispatch(fetchTodos()),
    setViewMode: (mode) => dispatch(setViewMode(mode)),
    toggleTodoMenu: (id) => dispatch(toggleTodoMenu(id))
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListView);
