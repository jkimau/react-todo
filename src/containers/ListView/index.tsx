import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import TodoRow from 'components/TodoRow';
import Loading from 'components/Loading';
import { mainBG, todoRowBorder } from 'global/colors';
import { fetchTodos, setViewMode, toggleTodoMenu } from 'actions';
// import { TodoPropTypes } from 'propTypes';

const ListViewWrapper = styled.div`
  height: 100%;
  padding-top: 40px;
  background: ${mainBG};
`;

const ContentContainer = styled.div`
  min-width: 400px;
  max-width: 590px;
  padding: 0 20px;
  margin: 0 auto;
  box-sizing: border-box;
`;

const TodoListWarpper = styled.div`
  border: 1px solid ${todoRowBorder};
`;

export interface ListViewPropTypes {
  todos: Array<{
    id: number,
    title: string,
    date: string,
    details: string,
    completed: boolean,
    menuOpen: boolean
  }>,
  isFetching: boolean,
  fetchTodos(): void,
  setViewMode(mode: string): void,
  toggleTodoMenu(id: number): void
}

export class ListView extends Component<ListViewPropTypes, null> {
  componentDidMount() {
    if (!this.props.isFetching) {
      this.props.fetchTodos();
    }
  }

  shouldComponentUpdate(nextState) {
    const hasSameTodos = this.props.todos === nextState.todos;
    return !hasSameTodos;
  }

  render() {
    const todoListWrapper = () => (
      <TodoListWarpper>
        {this.props.todos.map(todo => (
          <TodoRow
            key={todo.id}
            todo={todo}
            toggleTodoMenu={this.props.toggleTodoMenu}
          />
        ))}
      </TodoListWarpper>
    );

    const loading = () => (<Loading/>);

    return (
      <ListViewWrapper>
        <h1>List view page</h1>
        <ContentContainer>
          {
            this.props.todos
            && this.props.todos.length > 0
            ? todoListWrapper()
            : loading()
          }
        </ContentContainer>
      </ListViewWrapper>
    );
  }
}

const mapStateToProps = ({ todoState: { isFetching, todos }}) => {
  return { isFetching, todos }
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTodos: () => dispatch(fetchTodos()),
    setViewMode: (mode: string) => dispatch(setViewMode(mode)),
    toggleTodoMenu: (id: number) => dispatch(toggleTodoMenu(id))
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListView);
