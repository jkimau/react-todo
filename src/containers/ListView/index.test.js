import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { ListView } from '../ListView';
import TodoRow from 'components/TodoRow';
import Loading from 'components/Loading';

describe('<ListView/>', () => {
  const defaultProps = {
    fetchTodos: jest.fn(),
    setViewMode: jest.fn(),
    toggleTodoMenu: jest.fn(),
    isFetching: false,
    todos: []
  };

  const compileComponent = (props = defaultProps) => {
    return shallow(<ListView {...props}/>);
  };

  test('snapshot with no todos', () => {
    const tree = renderer.create(
      <ListView {...defaultProps}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })

  test('snapshot with todos', () => {
    const newProps = Object.assign({}, defaultProps);
    newProps.todos = [
      {
        id: 1,
        title: 'test 1',
        date: '2017-04-20T00:00:00Z',
        details: 'test details',
        completed: false,
        menuOpen: false
      }
    ];

    const tree = renderer.create(
      <ListView {...newProps}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should render <Loading/>', () => {
    const wrapper = compileComponent();

    expect(wrapper.find(Loading).length).toBe(1);
    expect(wrapper.find(TodoRow).length).toBe(0);
  })

  test('should render <TodoRow/>', () => {
    const newProps = Object.assign({}, defaultProps);
    newProps.todos = [
      {
        id: 1,
        title: 'test 1',
        date: '2017-04-20T00:00:00Z',
        details: 'test details',
        completed: false,
        menuOpen: false
      }
    ];
    const wrapper = compileComponent(newProps);

    expect(wrapper.find(TodoRow).length).toBe(1);
    expect(wrapper.find(Loading).length).toBe(0);
  });

  test('should pass correct props to <TodoRow/>', () => {
    const newProps = Object.assign({}, defaultProps);
    newProps.todos = [
      {
        id: 1,
        title: 'test 1',
        date: '2017-04-20T00:00:00Z',
        details: 'test details',
        completed: false,
        menuOpen: false
      }
    ];
    const wrapper = compileComponent(newProps);

    expect(wrapper.find(TodoRow).prop('todo')).toEqual({
      id: 1,
      title: 'test 1',
      date: '2017-04-20T00:00:00Z',
      details: 'test details',
      completed: false,
      menuOpen: false
    });
  });

  test('should call fetchTodos()', () => {
    const wrapper = mount(<ListView {...defaultProps}/>);

    expect(wrapper.instance().props.fetchTodos).toHaveBeenCalled();
  });

  test('should not call fetchTodos()', () => {
    const newProps = Object.assign({}, defaultProps);
    // has to set spy again as Object.assign only copies top level
    newProps.fetchTodos = jest.fn();
    newProps.isFetching = true;
    const wrapper = mount(<ListView {...newProps}/>);

    expect(wrapper.instance().props.fetchTodos).not.toHaveBeenCalled();
  });
});
