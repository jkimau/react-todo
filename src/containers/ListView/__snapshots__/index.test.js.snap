exports[`<ListView/> snapshot with no todos 1`] = `
<div
  className="jjZMrb">
  <h1>
    List view page
  </h1>
  <div
    className="bxhHXa">
    <div
      className="khmhuD">
      Loading...
    </div>
  </div>
</div>
`;

exports[`<ListView/> snapshot with todos 1`] = `
<div
  className="jjZMrb">
  <h1>
    List view page
  </h1>
  <div
    className="bxhHXa">
    <div
      className="cqDRAa">
      <div
        className="delrmX">
        <div
          className="Eszii">
          <span
            className="todo-row-header">
            test 1
          </span>
          <span
            className="todo-row-date">
            2017-04-20T00:00:00Z
          </span>
          <span
            className="todo-menu-trigger bPMTRm"
            onClick={[Function]}>
            open
          </span>
          <ul
            className="todo-row-menu slide-enter gHZgXS"
            onClick={[Function]}>
            <li>
              Mark complete
            </li>
            <li>
              Delete
            </li>
          </ul>
        </div>
        <div
          className="todo-row-details hAJdRI">
          test details
        </div>
      </div>
    </div>
  </div>
</div>
`;
